lib/srv/db/dynamodb/engine.go

```diff
@@ -0,0 +1,421 @@
+		Status:     http.StatusText(code),
+		StatusCode: code,
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		// ContentLength is the authoritative value in the response,
+		// no need to also add the "Content-Length" header (source: go doc http.Response.Header).
+		ContentLength: int64(len(body)),
+		Header: map[string][]string{
+			"Content-Type": {"application/json"},
+		},
+		Body: ioutil.NopCloser(bytes.NewBuffer(body)),
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.WithError(err).Error("failed to send error response to DynamoDB client")
+		return
+	}
+}
+
+// HandleConnection authorizes the incoming client connection, connects to the
+// target DynamoDB server and starts proxying requests between client/server.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) error {
+	err := e.checkAccess(ctx, e.sessionCtx)
+	e.Audit.OnSessionStart(e.Context, e.sessionCtx, err)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer e.Audit.OnSessionEnd(e.Context, e.sessionCtx)
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := e.process(ctx, req); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// process reads request from connected dynamodb client, processes the requests/responses and send data back
+// to the client.
+func (e *Engine) process(ctx context.Context, req *http.Request) error {
+	defer req.Body.Close()
+	body, err := io.ReadAll(io.LimitReader(req.Body, teleport.MaxHTTPRequestSize))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	service, err := e.getService(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	signingName, err := serviceToSigningName(service)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	uri, err := e.getTargetURI(service)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	roundTripper, err := e.getRoundTripper(ctx, uri)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// rewrite the request URL and headers before signing it.
+	reqCopy, err := rewriteRequest(ctx, req, uri, bytes.NewReader(body))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	region := e.sessionCtx.Database.GetAWS().Region
+	roleArn := libaws.BuildRoleARN(e.sessionCtx.DatabaseUser, region, e.sessionCtx.Database.GetAWS().AccountID)
+	signedReq, err := e.signingSvc.SignRequest(reqCopy,
+		&libaws.SigningCtx{
+			SigningName:   signingName,
+			SigningRegion: region,
+			Expiry:        e.sessionCtx.Identity.Expires,
+			SessionName:   e.sessionCtx.Identity.Username,
+			AWSRoleArn:    roleArn,
+		})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Send the request to DynamoDB API.
+	resp, err := roundTripper.RoundTrip(signedReq)
+	if err != nil {
+		// convert the error from round tripping to try to get a trace error.
+		err = common.ConvertConnectError(err, e.sessionCtx)
+		e.Log.WithError(err).Error("Request failed.")
```

Comment:
Do we log an audit event when the request fails anywhere? Feels we should capture at least what the client sends.

Reply:
I didn't consider that other engines emit the audit event regardless of roundtrip error

Reply:
most database engines capture the client requests for the audit and do not wait for the response. so I guess we care more about what the client sends than what the server responds.

Reply:
Updated to log an audit event for the request when roundtrip fails to get a response